src/compiler/emitter.ts

```diff
@@ -2777,34 +2776,61 @@ var __param = this.__param || function(index, decorator) { return function (targ
                 return getCombinedNodeFlags(node.parent);
             }
 
-            function renameNonTopLevelLetAndConst(node: Node): void {
+            function renameBlockScopedValueIfNecessary(node: Node, meaning: SymbolFlags): void {
+                // currently supported scenarios - classes and block scoped variables
+                Debug.assert(meaning === SymbolFlags.BlockScopedVariable || meaning === SymbolFlags.Class);
+
                 // do not rename if
                 // - language version is ES6+
                 // - node is synthesized
                 // - node is not identifier (can happen when tree is malformed)
-                // - node is definitely not name of variable declaration. 
-                // it still can be part of parameter declaration, this check will be done next
                 if (languageVersion >= ScriptTarget.ES6 ||
                     nodeIsSynthesized(node) ||
-                    node.kind !== SyntaxKind.Identifier ||
-                    (node.parent.kind !== SyntaxKind.VariableDeclaration && node.parent.kind !== SyntaxKind.BindingElement)) {
+                    node.kind !== SyntaxKind.Identifier) {
                     return;
                 }
 
-                let combinedFlags = getCombinedFlagsForIdentifier(<Identifier>node);
-                if (((combinedFlags & NodeFlags.BlockScoped) === 0) || combinedFlags & NodeFlags.Export) {
-                    // do not rename exported or non-block scoped variables
+                let flags: NodeFlags;
+                let container: Node;
+                if (meaning === SymbolFlags.BlockScopedVariable) {
+                    // block scoped variable case
+                    // do not rename node is definitely not name of variable declaration. 
+                    if (node.parent.kind !== SyntaxKind.VariableDeclaration && node.parent.kind !== SyntaxKind.BindingElement) {
+                        return;
+                    }
+
+                    flags = getCombinedFlagsForIdentifier(<Identifier>node);
+                    if (!(flags & NodeFlags.BlockScoped)) {
+                        // do not rename non-block scoped variables
+                        return;
+                    }
+
+                    let list = getAncestor(node, SyntaxKind.VariableDeclarationList);
+                    if (list.parent.kind === SyntaxKind.VariableStatement) {
+                        container = list.parent.parent;
+                    }
+                }
+                else {
+                    if (node.parent.kind !== SyntaxKind.ClassDeclaration) {

```

Comment:
Would it be simpler to just include Class in BlockScopedVariable? We may end up needing to do this anyway for deep class declarations.

Better yet, maybe have another symbol flag called BlockScoped that is the union of Class and BlockScopedVariable.